### PR TITLE
Explicit typed stream interfaces

### PR DESCRIPTION
**Pull request by bergos**

### CODE REVIEW

File: examples/forEach.ttl [hunk 1]
```diff
@@ -1,29 +1,46 @@
...
-    code:implementedBy [ a code:EcmaScript ;
-            code:link <node:fs#createReadStream> ] .
-
-<http://example.org/pipeline#serializeJson> code:implementedBy [ a code:EcmaScript ;
-            code:link <file:steps/serializeJson.js> ] .
-
-<http://example.org/pipeline#subPipeline> a p:Pipeline ;
-    p:steps [ p:stepList ( <http://example.org/pipeline#duplicate> ) ] .
-
+<#pipeline> a p:Pipeline, p:Readable ;
```
Comment (tpluscode) [inline examples/forEach.ttl:5]: Wish you didn't reformat this one completely. The relevant changes are to add `p:Readable` etc?

File: test/loader/pipeline.test.js [hunk 2]
```diff
@@ -30,7 +30,7 @@ describe('pipeline loader', () => {
 
     // when
     const pipeline = loader(node, dataset, { context, variables, basePath: context.basePath })
-    await pipeline.initVariables()
+    await pipeline._pipeline.initVariables()
```
Comment (tpluscode) [inline test/loader/pipeline.test.js:33]: Is this something clients would be typing as well, to get the inner `_pipeline` object?

File: test/createBaseStream.test.js [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+      const pipeline = { writable: true }
+      const init = () => {}
+      const write = () => {
+        touched = true
+      }
+      const stream = createBaseStream(pipeline, { init, write })
+
+      // when
+      stream.write('')
+      await sleep(10)
```
Comment (tpluscode) [inline test/createBaseStream.test.js]: Why sleep and not stream events or something?

File: lib/pipeline.js [hunk 4]
```diff
@@ -1,29 +1,42 @@
...
     this.context.log = new Logger(this.node, { master: context.log })
 
-    this.init = once(() => this._init().catch(err => this.emit('error', err)))
-    this.initVariables = once(async () => {
-      const parsedVariables = await this.parseVariables()
-      this.variables = new Map([...parsedVariables, ...this.variables])
-      this.context.variables = this.variables
-    })
+    this.stream = createPipelineStream(this)
+    this.context.pipeline = this.stream
```
Comment (tpluscode) [inline lib/pipeline.js]: Hm, this is suspicious. Before `context.pipeline` was assigned the pipeline itself and now it's a stream?

File: examples/forEach.ttl (hunk 1 above)
Comment (bergos) [reply examples/forEach.ttl:5]: Sorry that I used the same commit, but for an example it was just not readable enough. Jup, just added the `p:Readable` and `p:ReadableObjectMode, p:WritableObjectMode` to the sub-pipeline.

File: test/loader/pipeline.test.js (hunk 2 above)
Comment (bergos) [reply test/loader/pipeline.test.js:33]: I just added `._pipeline` to have access to the inner object during debugging. I think that test should be even changed to init the variables by calling `.read()`, which will also trigger `.init()` + `.initVariables()`.

File: test/createBaseStream.test.js (hunk 3 above)
Comment (bergos) [reply test/createBaseStream.test.js]: Could be used. The idea was to not rely on anything besides the event loop.

File: lib/pipeline.js (hunk 4 above)
Comment (bergos) [reply lib/pipeline.js]: I understand and I think it deserves a comment, but it's right. Before `context.pipeline` was assigned with `this`, cause there was only one pipeline object which represented the inner and outer pipeline. Now the stream is the outer/public interface. As the context should be used by other packages, the public interface should be assigned.

Comment (tpluscode) [reply lib/pipeline.js]: Ok, makes sense. Should we still call this `context.pipeline`?

File: test/createBaseStream.test.js (hunk 3 above)
Comment (tpluscode) [reply test/createBaseStream.test.js]: Well, I have been conditioned to consider any artificial delays in test code to be a bit of a smell. If there is inherent asynchronicity in the test code then I think we should reflect that in the tests too

File: test/loader/pipeline.test.js (hunk 2 above)
Comment (tpluscode) [reply test/loader/pipeline.test.js:33]: I think we can live with that, if this is only for debugging or forcing the initial state in test

**Pull request merged**